probe-rs/src/probe/blackmagic/arm.rs

```diff
@@ -0,0 +1,840 @@
+                data,
+            },
+            ProtocolVersion::V3 => RemoteCommand::MemWriteV3 {
+                index: self.index,
+                apsel: self.apsel,
+                csw: self.csw,
+                align,
+                offset: offset
+                    .try_into()
+                    .map_err(|_| ArmError::AddressOutOf32BitAddressSpace)?,
+                data,
+            },
+            ProtocolVersion::V4 => RemoteCommand::MemWriteV4 {
+                index: self.index,
+                apsel: self.apsel,
+                csw: self.csw,
+                align,
+                offset,
+                data,
+            },
+        };
+        let result = self
+            .probe
+            .probe
+            .command(command)
+            .map_err(|e| ArmError::Probe(e.into()))?
+            .0;
+        if result == 0 {
+            Ok(())
+        } else {
+            Err(ArmError::Probe(DebugProbeError::Other(format!(
+                "probe returned unexpected result: {}",
+                result
+            ))))
+        }
+    }
+
+    fn write(&mut self, align: u8, offset: u64, data: &[u8]) -> Result<(), ArmError> {
+        let chunk_size = super::BLACK_MAGIC_REMOTE_SIZE_MAX / 2 - 40;
+        for (chunk_index, chunk) in data.chunks(chunk_size).enumerate() {
+            self.write_slice(
+                align,
+                chunk_index as u64 * chunk_size as u64 + offset,
+                chunk,
+            )?;
+        }
+        Ok(())
+    }
+}
+
+impl<R: Read + Send + core::fmt::Debug + 'static, W: Write + Send + core::fmt::Debug + 'static>
+    MemoryInterface<ArmError> for BlackMagicProbeMemoryInterface<'_, R, W>
+{
+    fn supports_native_64bit_access(&mut self) -> bool {
+        self.probe.probe.remote_protocol == ProtocolVersion::V4
+    }
+
+    fn read_64(&mut self, address: u64, data: &mut [u64]) -> Result<(), ArmError> {
+        self.read(address, data.as_bytes_mut())
+    }
+
+    fn read_32(&mut self, address: u64, data: &mut [u32]) -> Result<(), ArmError> {
+        self.read(address, data.as_bytes_mut())
+    }
+
+    fn read_16(&mut self, address: u64, data: &mut [u16]) -> Result<(), ArmError> {
+        self.read(address, data.as_bytes_mut())
+    }
+
+    fn read_8(&mut self, address: u64, data: &mut [u8]) -> Result<(), ArmError> {
+        self.read(address, data.as_bytes_mut())
+    }
+
+    fn write_64(&mut self, address: u64, data: &[u64]) -> Result<(), ArmError> {
+        self.write(3, address, data.as_bytes())
+    }
+
+    fn write_32(&mut self, address: u64, data: &[u32]) -> Result<(), ArmError> {
+        self.write(2, address, data.as_bytes())
+    }
+
+    fn write_16(&mut self, address: u64, data: &[u16]) -> Result<(), ArmError> {
+        self.write(1, address, data.as_bytes())
+    }
+
+    fn write_8(&mut self, address: u64, data: &[u8]) -> Result<(), ArmError> {
+        self.write(0, address, data.as_bytes())
```

Comment:
Can we avoid magic numbers here? `align = 0` doesn't have much meaning to me, I'm guessing this is some device-specific constant for 8-bit alignment. I'd prefer an `enum` that you can just `as u8` into the numeric value.

Reply:
Good idea, I'll do that.